### ISSUE TITLE
Immutable Optimizer Updates and Training Loops [AI/ML]

### ISSUE THREAD

**Issue by mcoughlin** [cool-japan/scirs#94]
**Describe the AI/ML feature you'd like to see**
I would like to see support for immutable optimizer-style parameter updates and reusable training loops in scirs2-autograd, so that users could build stable training or variational inference loops without mutating tensors in-place or leaking computation graphs. I really struggled while trying to make a SVI framework happen naturally today based on the existing examples in scirs2-autograd.

**Is this feature available in existing ML libraries?**
PyTorch (torch.optim, loss.backward(), param = param - lr * grad), JAX (optax, functional parameter updates), Pyro / NumPyro all treat parameters as immutable values and return new versions at each step.

**How would this feature benefit SciRS2 users?**
Generally speaking, this would make development better. At least I saw: Graph memory leaks (“Too many tensors in this graph”), input_mut panics, a lack of a public way to reset or reuse graphs between iterations, etc.. And of course this would allow users to work on variational inference problems.

**Implementation ideas (optional)**
I suspect it wouldn't be hard to expose a public Context::clear() / Graph::clear() to safely reuse graphs between iterations. Harder might be some kind of functional optimizer that returns new tensors instead of mutating inputs:
e.g. let (new_params) = sgd.step(params, grads, ctx);. On a related note, some kind of "training loop safe mode" that automatically prunes old graph nodes.

**Additional context**
I was working on a standard ELBO optimization loop and struggling. Of course, this is at least partially my lack of familiarity with the package, but I did feel like I was fighting the implementation more than I expected.
